### PR TITLE
[GH-109] Update the meeting post and send DM to host when users joined meeting before host

### LINKED ISSUE #109: Zoom: Alert the channel when a particpant has joined a meeting before host
There is a webhook event notification available for this scenario in Zoom's API.

Users would like to know if there are others who have joined the meeting in the channel.  We should post a message to the channel about the state like "Someone joined the Zoom meeting before the host had a chance to join...."

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
#### Summary
- Update the meeting post and send a DM to the host when users join the meeting before the host.

#### Screenshots
Custom post in channel:
image

Slack attachment in channel:
image
DM message:
image
#### What to test?
- The host is getting a DM from the Zoom bot when the user(s) joined the meeting before him.
- The meeting post is updated in the channel if the users join the meeting before the host.

Note: Check for custom posts and Slack attachments (Slack attachments if displayed in the mobile app or when the plugin is disabled).

###### Steps to reproduce:
Note: Ensure you have added the "Participant was waiting for host to join", "Participant joined meeting before host" and "Participant/Host joined meeting" webhook events on your Zoom app.
1. Connect your Zoom account.
2. Create a meeting in any channel. (Do not join the meeting)
3. Join the meeting using another MM user.

#### Ticket Link
Fixes #109

### CODE REVIEW

**Comment by mickmister** [comment]
@matthewbirtch What are your thoughts on this feature? I'm thinking we should make this a configurable user setting, though not sure if it should be opt-in or opt-out

**Comment by matthewbirtch** [comment]
@mickmister I think it's a cool idea - I agree that we should provide a way to opt out of this in the zoom plugin settings. "For meetings I host, notify me when participants join before me".

I like the "X participants waiting for the host", but I'm not sure we need the "Host is already in the meeting" added to the Zoom meeting post. Seems unnecessary to me.

**Comment by raghavaggarwal2308** [comment]
@mickmister Just for clarification:
1. We need to add a per-user setting (like in the `/zoom setting` command) to disable the DM notification. (Default value : true)
2. We need to remove "Host is already in the meeting" displayed on the meeting post.

**Comment by mickmister** [comment]
@raghavaggarwal2308 Yes sounds good :+1:  Can we also change "participants(s)" to adapt as "participant" and "participants"?

**Comment by raghavaggarwal2308** [comment]
> @mickmister Just for clarification:
> 
> 1. We need to add a per-user setting (like in the `/zoom setting` command) to disable the DM notification. (Default value : true)
> 2. We need to remove "Host is already in the meeting" displayed on the meeting post.

@mickmister @matthewbirtch Updated the PR with these changes

**Comment by mickmister** [comment]
Also there are some merge conflicts to resolve here

**Comment by matthewbirtch** [comment]
what's the best way for me to test the ux here @mickmister?

File: server/zoom/webhook.go [hunk 1]
```diff
@@ -15,21 +15,27 @@ const (
...
 	RecentlyCreated              = "RECENTLY_CREATED"
 
-	EventTypeMeetingStarted  EventType = "meeting.started"
-	EventTypeMeetingEnded    EventType = "meeting.ended"
-	EventTypeValidateWebhook EventType = "endpoint.url_validation"
+	EventTypeMeetingStarted           EventType = "meeting.started"
+	EventTypeMeetingEnded             EventType = "meeting.ended"
+	EventTypeValidateWebhook          EventType = "endpoint.url_validation"
+	EventTypeParticipantJoined        EventType = "meeting.participant_joined"
+	EventTypeParticipantJoinedWaiting EventType = "meeting.participant_jbh_waiting"
```
[the remark this answers is not in the document]
Comment (Kshitij-Katiyar) [reply server/zoom/webhook.go:22]: join before host, Its the name of the webhook event from the zoom side @mickmister

File: server/webhook.go [hunk 2]
```diff
@@ -134,6 +136,98 @@ func (p *Plugin) handleMeetingEnded(w http.ResponseWriter, r *http.Request, body
...
+				p.API.LogWarn("Could not get direct channel", "err", appErr)
+				http.Error(w, appErr.Error(), appErr.StatusCode)
+				return
+			}
+
+			_, appErr = p.API.CreatePost(&model.Post{
+				ChannelId: channel.Id,
+				Message:   fmt.Sprintf("**%s** has joined the [meeting](%s) before you.", participant.UserName, meetingURL),
+				UserId:    p.botUserID,
+			})
```
[the remark this answers is not in the document]
Comment (Kshitij-Katiyar) [reply server/webhook.go:202]: @mickmister We are getting the webhook payload every time somebody joins. One thing could be that we can update the initial DM only but for that, we have to save the DM post ID somewhere maybe KV store. But atm it maybe overkill.

[the remark this answers is not in the document]
Comment (mickmister) [reply server/webhook.go:202]: Yeah this is a UX question for sure. @matthewbirtch If multiple join before the host joins, should we send multiple DMs, or change anything about the original DM sent to the host, when a second person joins?

File: server/webhook.go [hunk 3]
```diff
@@ -134,6 +136,98 @@ func (p *Plugin) handleMeetingEnded(w http.ResponseWriter, r *http.Request, body
...
+	}
+
+	post.Props["meeting_host_joined"] = isHostJoined
+	post.Props["meeting_waiting_count"] = waitingCount
+
+	postAttachments := post.Attachments()
+	postAttachments[0].Text = fmt.Sprintf("Meeting ID: [%s](%s)\n\n%s\n\n[Join Meeting](%s)", meetingID, meetingURL, participantsJoinedMsg, meetingURL)
+	model.ParseSlackAttachment(post, postAttachments)
+
+	_, appErr = p.API.UpdatePost(post)
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/webhook.go:218]: @mickmister We will have to handle that logic in the meeting ended webhook. Zoom does not send any field for numberOfParticipantsJoined in the [meeting ended webhook](https://developers.zoom.us/docs/meeting-sdk/webhooks/#operation/meeting.ended). I think we will have to make an extra API call to get that data. Maybe we can create another issue for this if you want?

**Comment by raghavaggarwal2308** [comment]
@mickmister Fixed the review comments added by you. Also, I have setup the zoom plugin on a cloud server. Please refer [this](https://hub.mattermost.com/partners/pl/exocwmmkm3g1inpjwr886y4eey)

**Comment by matthewbirtch** [comment]
@raghavaggarwal2308 I've got things working properly, but a few notes from my testing:
- Joining Zoom via browser seems to be updating the post, but joining via desktop app does not
- Once I join and then leave, the post doesn't get updated - still says 1 participant waiting for host. 
- If I have >1 participant joining, that also does not seem to be reflected properly in the post.  It still says '1 participant waiting for host'
- Also, if a second participant joins, I don't seem to be getting notified in a DM
- Can we make some text changes:
    - `1 participant is waiting for the host to join` (don't use `participant(s)`)
    - `X participants are waiting for the host to join` (don't use `participant(s)`)
    -  `The host has joined the meeting`

**Comment by raghavaggarwal2308** [comment]
@matthewbirtch Few replies/clarification from above comment:
1. The post should update whenever a user joins the meeting, not sure why its not updating for you.
2. The host get notified everytime a user joined the meeting before him. 
- There must be some issue because of which you are not getting the webhook from Zoom on joining with second user and hence you are not getting notified and post is not getting updated. I will see If I can reproduce the issue
- Also, @mickmister had a suggesting about multiple DMs [here](https://github.com/mattermost/mattermost-plugin-zoom/pull/334#discussion_r1600598698)

3. `The host has joined the meeting`: We added the logic to show a similar message when we created the PR but it was remove later because of [this](https://github.com/mattermost/mattermost-plugin-zoom/pull/334#issuecomment-1946824524) comment. Please let us know if we should add it back.

**Comment by matthewbirtch** [comment]
> * Also, @mickmister had a suggesting about multiple DMs [here](https://github.com/mattermost/mattermost-plugin-zoom/pull/334#discussion_r1600598698)

I'm concerned about notifying for every single participant that joins - especially for large meetings. I like the idea @mickmister had of updating the original DM message with the participants that have joined. That would be a way to avoid multiple notifications. That would need to max out after more 3 participants join. For example, 
- `Jim and Alice have joined the meeting before you`
- `Jim, Alice, and Frank have joined the meeting before you`
- `Jim, Alice, Frank, and X more have joined the meeting before you`.

> `The host has joined the meeting`: We added the logic to show a similar message when we created the PR but it was remove later because of [this](https://github.com/mattermost/mattermost-plugin-zoom/pull/334#issuecomment-1946824524) comment. Please let us know if we should add it back.

Thanks @raghavaggarwal2308 - I completely forgot I made that comment. I'm aligned to the original comment, no need to bring it back. Could you remove this from the PR description?

**Comment by mickmister** [comment]
I'm thinking we'll want a clear way for users to opt-out of this feature. Also note that this was a suggestion from a previous PM and not direct ask from a customer. @raghavaggarwal2308 What do you think about the remaining work and potential maintenance costs of the feature?

**Comment by raghavaggarwal2308** [comment]
@wiggin77 Do you think we should continue our work on this PR?

**Comment by raghavaggarwal2308** [comment]
Closing for now as unplanned, can reopen if required in future

**Pull request closed without merging**